### PR TITLE
Better wildcards for COMPLUS_JIT variables

### PR DESCRIPTION
**Pull request by briansull**
With this change we can use a prefix and wildcard match
This allows us to match classes and methods that start with a prefix.
Example:
1. My*:Method would match a method named "Method" inside any class that starts with "My"
2. \*:get_* or simply get_* both will match any method that starts with "get_"

This allows us to match some Managed C++ methods that we could not otherwise match:

By using COMPLUS_JitDump=std._Find_if*
we can match this managed C++ method:

 <Module>:std._Find_if<class std::_Tree_unchecked_iterator<class std::_Tree_val<struct std::_Tree_simple_types<struct std::pair<struct TIP_MAP_KEY const ,struct TIP_MAP_VALUE> > > >,struct TIP_FIND_ALG>(long,struct,struct,struct)

### CODE REVIEW

File: src/jit/jitconfig.cpp [hunk 1]
```diff
@@ -239,9 +300,21 @@ void JitConfigValues::MethodSet::destroy(ICorJitHost* host)
     m_names = nullptr;
 }
 
-static bool matchesName(const char* const name, int nameLen, const char* const s2)
+static bool matchesName(const char* const name, int nameLen, bool wildcardAtEnd, const char* const s2)
 {
-    return strncmp(name, s2, nameLen) == 0 && s2[nameLen] == '\0';
+    // s2 must start with the characters in 'name'
+    bool result = (strncmp(name, s2, nameLen) == 0);
```
[the remark this answers is not in the document]
Comment (briansull) [reply src/jit/jitconfig.cpp]: At one point I had written return result, so you are correct it is unnecessary now.

**Comment by briansull** [comment]
I tested all of the interesting cases by hand.

When I broke things, there were a handful of dump re;ated test failures I saw on x86 desktop:

```
jit-opt-DumpDisasm-RuntimeTypeHandleOpt-JITOpt	Deterministic 1:DifferFromBaseline	FAIL, FAIL	PASS
jit-Opt-DumpDisasm-JitMinOpts-TestLVRefCnt	Deterministic 1:DifferFromBaseline	FAIL, FAIL	PASS
jit-Opt-DumpDisasm-JitMinOpts-TestBBCnt		Deterministic 1:DifferFromBaseline	FAIL, FAIL	PASS
jit-Opt-DumpDisasm-JitMinOpts-TestB189877	Deterministic 1:DifferFromBaseline	FAIL, FAIL	PASS
jit-Opt-DumpDisasm-JitMinOpts-TestCodeSize	Deterministic 1:DifferFromBaseline	FAIL, FAIL	PASS
jit-Opt-DumpDisasm-JitMinOpts-TestLVNumCnt	Deterministic 1:DifferFromBaseline	FAIL, FAIL	PASS
jit-Opt-DumpDisasm-JitMinOpts-TestInstrCnt	Deterministic 1:DifferFromBaseline	FAIL, FAIL	PASS

```

**Comment by CarolEidt** [comment]
@briansull - thanks for doing this! Could you update the documentation here: https://github.com/dotnet/coreclr/blob/master/Documentation/botr/ryujit-overview.md#how-to-create-a-jitdump ?

**Comment by briansull** [comment]
Carol, I will add a section on the set of string matching patterns supported by COMPLUS_JitDump, etc...

**Comment by CarolEidt** [comment]
@briansull - that would be great, thanks!

**Pull request merged**